apollo-router/src/cache/mod.rs

```diff
@@ -0,0 +1,178 @@
+                    let mut locked_wait_map = wait_map.lock().await;
+                    let _ = locked_wait_map.remove(&k);
+                });
+
+                Entry {
+                    inner: EntryInner::First {
+                        sender,
+                        key: key.clone(),
+                        cache: self.clone(),
+                        _drop_signal,
+                    },
+                }
+            }
+        }
+    }
+
+    pub(crate) async fn insert(&self, key: K, value: V) {
+        self.storage.insert(key, value.clone()).await;
+    }
+
+    pub(crate) async fn remove_wait(&self, key: &K) {
+        let mut locked_wait_map = self.wait_map.lock().await;
+        let _ = locked_wait_map.remove(key);
+    }
+}
+
+pub(crate) struct Entry<K: Clone + Send + Eq + Hash, V: Clone + Send> {
+    inner: EntryInner<K, V>,
+}
+enum EntryInner<K: Clone + Send + Eq + Hash, V: Clone + Send> {
+    First {
+        key: K,
+        sender: broadcast::Sender<V>,
+        cache: DeduplicatingCache<K, V>,
+        _drop_signal: oneshot::Sender<()>,
+    },
+    Receiver {
+        receiver: broadcast::Receiver<V>,
+    },
+    Value(V),
+}
+
+#[derive(Debug)]
+pub(crate) enum EntryError {
+    Closed,
+    IsFirst,
+}
+
+impl<K, V> Entry<K, V>
+where
+    K: Clone + Send + Eq + Hash + 'static,
+    V: Clone + Send + 'static,
+{
+    pub(crate) fn is_first(&self) -> bool {
+        matches!(self.inner, EntryInner::First { .. })
+    }
+
+    pub(crate) async fn get(self) -> Result<V, EntryError> {
+        match self.inner {
+            // there was already a value in cache
+            EntryInner::Value(v) => Ok(v),
+            EntryInner::Receiver { mut receiver } => {
+                receiver.recv().await.map_err(|_| EntryError::Closed)
+            }
+            _ => Err(EntryError::IsFirst),
+        }
+    }
+
+    pub(crate) async fn insert(self, value: V) {
+        if let EntryInner::First {
+            key,
+            sender,
+            cache,
+            _drop_signal,
+        } = self.inner
+        {
+            cache.insert(key.clone(), value.clone()).await;
+            cache.remove_wait(&key).await;
+            let _ = sender.send(value);
+        }
+    }
+
+    /*
+    /// sends the value without storing it into the cache
+    pub(crate) async fn send(self, value: V) {
+        if let EntryInner::First {
+            sender, cache, key, ..
+        } = self.inner
+        {
+            let _ = sender.send(value);
+            cache.remove_wait(&key).await;
+        }
+    }*/
+}
+
+/*
+async fn example_cache_usage(k: String, cache: &mut DedupCache<String, String>) -> String {
+    let entry = cache.get(&k).await;
+
+    if entry.is_first() {
+        // potentially long and complex async task that can fail
+        let value = "hello".to_string();
+        entry.insert(value.clone()).await;
+        value
+    } else {
+        entry.get().await.unwrap()
+    }
+}
+*/
```

Comment:
the `send()` function will be used in a follow up PR. It's useful when we want to return the result of the computation to all the waiting queries, but without storing the value in the cache (example: failed network call because the subgraph was unreachable). Unfortunately, the linter complains if I leave some `pub(crate)` code but do not use it. Should I tag it with `#[allow(unused)] instead?

The example function should be transformed into a unit test.

Reply:
Yes, I'd prefer `allow(unused)` with an explanation comment over commented out code

Reply:
makes sense. I changed that in 1d765c9d